File: start/news-reader-ios/NewsReader/presentation/MainViewModel.swift
//
//  MainViewModel.swift
//  NewsReader
//
//  Created by marco.gomiero on 23.11.23.
//

import Foundation

@MainActor
class MainViewModel: ObservableObject {
    
    @Published
    var uiState: UiState = UiState.loading
    
    private var newsRepository: NewsRepository
    
    init(newsRepository: NewsRepository) {
        self.newsRepository = newsRepository
    }
    
    func loadNews() async {
        do {
            let newsList = try await newsRepository.fetchNews()
            uiState = UiState.success(newsList: newsList)
        } catch {
            print(error)
            uiState = UiState.error(message: error.localizedDescription)
        }
    }
}

File: start/news-reader-android/app/src/main/java/com/example/newsreader/presentation/MainViewModel.kt
package com.example.newsreader.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.newsreader.domain.News
import com.example.newsreader.domain.NewsRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

internal class MainViewModel(
    private val newsRepository: NewsRepository = NewsRepository()
) : ViewModel() {

    private val mutableUiState = MutableStateFlow<UiState>(UiState.Loading)
    val uiState: StateFlow<UiState> = mutableUiState.asStateFlow()

    init {
        viewModelScope.launch {
            try {
                val newsList = newsRepository.fetchNews()
                mutableUiState.update {
                    UiState.Success(newsList)
                }
            } catch (e: Exception) {
                mutableUiState.update {
                    UiState.Error(message = "Something went wrong")
                }
            }
        }
    }
}